{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import h5py\n",
    "import glob\n",
    "import datetime\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import logging\n",
    "import argparse\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import sib_ldsc_z as ld"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ldscpath = \"/disk/genetics/ukb/alextisyoung/haplotypes/relatives/bedfiles/ldscores/*[0-9].l2.ldscore.gz\"\n",
    "filenames = \"/disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_*.hdf5\"\n",
    "ldcolnames = [\"CHR\", \"SNP\", \"BP\", \"L2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ldscores= ld.read_ldscores(ldscpath, ldcolnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BP</th>\n",
       "      <th>L2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>752721</td>\n",
       "      <td>3.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs12184325</td>\n",
       "      <td>754105</td>\n",
       "      <td>2.113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs3131962</td>\n",
       "      <td>756604</td>\n",
       "      <td>3.793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs114525117</td>\n",
       "      <td>759036</td>\n",
       "      <td>2.103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs115991721</td>\n",
       "      <td>767096</td>\n",
       "      <td>1.103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHR          SNP      BP     L2\n",
       "0   1    rs3131972  752721  3.691\n",
       "1   1   rs12184325  754105  2.113\n",
       "2   1    rs3131962  756604  3.793\n",
       "3   1  rs114525117  759036  2.103\n",
       "4   1  rs115991721  767096  1.103"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ldscores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[b'6', b'rs11757325', b'0.269301865963754', b'203397', b'C', b'T'],\n",
       "       [b'6', b'rs80014302', b'0.269663095780137', b'203722', b'T', b'G'],\n",
       "       [b'6', b'rs80324517', b'0.270023385241025', b'204031', b'G', b'A'],\n",
       "       ...,\n",
       "       [b'6', b'rs17860838', b'198.250630831501', b'170889829', b'C',\n",
       "        b'T'],\n",
       "       [b'6', b'rs12206696', b'198.251798897992', b'170899075', b'T',\n",
       "        b'C'],\n",
       "       [b'6', b'rs7775031', b'198.254831357579', b'170907734', b'G',\n",
       "        b'A']], dtype='|S17')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_8.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_16.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_12.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_4.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_1.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_5.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_9.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_17.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_13.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_22.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_3.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_19.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_7.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_15.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_11.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_20.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_14.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_10.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_21.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_2.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_18.hdf5\n",
      "Reading in file:  /disk/genetics/ukb/alextisyoung/haplotypes/relatives/traits/9/chr_6.hdf5\n"
     ]
    }
   ],
   "source": [
    "files = glob.glob(filenames)\n",
    "\n",
    "file = files[0]\n",
    "print(\"Reading in file: \", file)\n",
    "hf = h5py.File(file, 'r')\n",
    "metadata = hf.get(\"bim\")[()]\n",
    "chromosome = metadata[:, 0]\n",
    "snp = metadata[:, 1]\n",
    "bp = metadata[:, 3]\n",
    "theta  = hf.get('estimate')[()]\n",
    "se  = hf.get('estimate_ses')[()]\n",
    "N = hf.get('N_L')[()]\n",
    "S = hf.get('estimate_covariance')[()]\n",
    "f = hf.get('freqs')[()]\n",
    "\n",
    "# normalizing S\n",
    "sigma2 = hf.get('sigma2')[()]\n",
    "tau = hf.get('tau')[()]\n",
    "phvar = sigma2+sigma2/tau\n",
    "\n",
    "if len(files) > 1:\n",
    "    for file in files[1:]:\n",
    "        print(\"Reading in file: \", file)\n",
    "        hf = h5py.File(file, 'r')\n",
    "        metadata = hf.get(\"bim\")[()]\n",
    "        chromosome_file = metadata[:, 0]  \n",
    "        snp_file = metadata[:, 1]\n",
    "        bp_file = metadata[:, 3]\n",
    "        theta_file  = hf.get('estimate')[()]\n",
    "        se_file  = hf.get('estimate_ses')[()]\n",
    "        S_file = hf.get('estimate_covariance')[()]\n",
    "        f_file = hf.get('freqs')[()]\n",
    "        N_file = hf.get('N_L')[()]\n",
    "\n",
    "        # normalizing S\n",
    "        sigma2 = hf.get('sigma2')[()]\n",
    "        tau = hf.get('tau')[()]\n",
    "\n",
    "        chromosome = np.append(chromosome, chromosome_file, axis = 0)\n",
    "        snp = np.append(snp, snp_file, axis = 0)\n",
    "        bp = np.append(bp, bp_file, axis = 0)\n",
    "        theta = np.append(theta, theta_file, axis = 0)\n",
    "        se = np.append(se, se_file, axis = 0)\n",
    "        S = np.append(S, S_file, axis = 0)\n",
    "        f = np.append(f, f_file, axis = 0)\n",
    "        N = np.append(N, N_file, axis = 0)\n",
    "\n",
    "# Constructing dataframe of data\n",
    "zdata = pd.DataFrame({'CHR' : chromosome,\n",
    "                    'SNP' : snp,\n",
    "                    'BP' : bp,\n",
    "                    'N' : N,\n",
    "                    \"f\" : f,\n",
    "                    'theta' : theta.tolist(),\n",
    "                    'se' : se.tolist(),\n",
    "                    \"S\" : S.tolist()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>N</th>\n",
       "      <th>f</th>\n",
       "      <th>theta</th>\n",
       "      <th>se</th>\n",
       "      <th>S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b'8'</td>\n",
       "      <td>b'rs11780869'</td>\n",
       "      <td>43342</td>\n",
       "      <td>0.075331</td>\n",
       "      <td>[-0.012398325838148594, 0.044778984040021896, ...</td>\n",
       "      <td>[0.02189680002629757, 0.039695702493190765, 0....</td>\n",
       "      <td>[[0.00047946989070624113, -0.00037018390139564...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b'8'</td>\n",
       "      <td>b'rs2003497'</td>\n",
       "      <td>43342</td>\n",
       "      <td>0.621533</td>\n",
       "      <td>[0.0010558661306276917, 0.0021548783406615257,...</td>\n",
       "      <td>[0.011948716826736927, 0.02137044072151184, 0....</td>\n",
       "      <td>[[0.000142771823448129, -0.0001093009341275319...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b'8'</td>\n",
       "      <td>b'rs56080987'</td>\n",
       "      <td>43342</td>\n",
       "      <td>0.980838</td>\n",
       "      <td>[0.043102409690618515, -0.06967756897211075, -...</td>\n",
       "      <td>[0.042830925434827805, 0.07517721503973007, 0....</td>\n",
       "      <td>[[0.0018344882410019636, -0.001305687241256237...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b'8'</td>\n",
       "      <td>b'rs116844674'</td>\n",
       "      <td>43342</td>\n",
       "      <td>0.962588</td>\n",
       "      <td>[0.04455234110355377, -0.06375543773174286, 0....</td>\n",
       "      <td>[0.03009759448468685, 0.05230757221579552, 0.0...</td>\n",
       "      <td>[[0.000905865163076669, -0.0006721090176142752...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b'8'</td>\n",
       "      <td>b'rs10488368'</td>\n",
       "      <td>43342</td>\n",
       "      <td>0.906026</td>\n",
       "      <td>[0.004531058948487043, -0.018941061571240425, ...</td>\n",
       "      <td>[0.01970919780433178, 0.03549977391958237, 0.0...</td>\n",
       "      <td>[[0.0003884524921886623, -0.000290763855446130...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR             SNP      N         f  \\\n",
       "0  b'8'   b'rs11780869'  43342  0.075331   \n",
       "1  b'8'    b'rs2003497'  43342  0.621533   \n",
       "2  b'8'   b'rs56080987'  43342  0.980838   \n",
       "3  b'8'  b'rs116844674'  43342  0.962588   \n",
       "4  b'8'   b'rs10488368'  43342  0.906026   \n",
       "\n",
       "                                               theta  \\\n",
       "0  [-0.012398325838148594, 0.044778984040021896, ...   \n",
       "1  [0.0010558661306276917, 0.0021548783406615257,...   \n",
       "2  [0.043102409690618515, -0.06967756897211075, -...   \n",
       "3  [0.04455234110355377, -0.06375543773174286, 0....   \n",
       "4  [0.004531058948487043, -0.018941061571240425, ...   \n",
       "\n",
       "                                                  se  \\\n",
       "0  [0.02189680002629757, 0.039695702493190765, 0....   \n",
       "1  [0.011948716826736927, 0.02137044072151184, 0....   \n",
       "2  [0.042830925434827805, 0.07517721503973007, 0....   \n",
       "3  [0.03009759448468685, 0.05230757221579552, 0.0...   \n",
       "4  [0.01970919780433178, 0.03549977391958237, 0.0...   \n",
       "\n",
       "                                                   S  \n",
       "0  [[0.00047946989070624113, -0.00037018390139564...  \n",
       "1  [[0.000142771823448129, -0.0001093009341275319...  \n",
       "2  [[0.0018344882410019636, -0.001305687241256237...  \n",
       "3  [[0.000905865163076669, -0.0006721090176142752...  \n",
       "4  [[0.0003884524921886623, -0.000290763855446130...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zdata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>N</th>\n",
       "      <th>f</th>\n",
       "      <th>theta</th>\n",
       "      <th>se</th>\n",
       "      <th>S</th>\n",
       "      <th>BP</th>\n",
       "      <th>L2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [CHR, SNP, N, f, theta, se, S, BP, L2]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zdata.merge(ldscores, how = \"inner\", on = [\"CHR\", \"BP\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
